File: plugins/OrchidPosts/src/main/java/com/eden/orchid/posts/permalink/pathTypes/SlugPathType.java
package com.eden.orchid.posts.permalink.pathTypes;

import com.eden.common.util.EdenUtils;
import com.eden.orchid.api.theme.pages.OrchidPage;
import com.eden.orchid.posts.PostsGenerator;
import com.eden.orchid.posts.PostsUtils;
import com.eden.orchid.posts.pages.PostArchivePage;
import com.eden.orchid.posts.pages.PostPage;
import com.eden.orchid.posts.permalink.PermalinkPathType;

import javax.inject.Inject;
import java.util.regex.Matcher;

public class SlugPathType extends PermalinkPathType {

    @Inject
    public SlugPathType() {
        super(100);
    }

    @Override
    public boolean acceptsKey(OrchidPage post, String key) {
        return key.equals("slug");
    }

    @Override
    public String format(OrchidPage page, String key) {
        if(page instanceof PostPage) {
            PostPage post = (PostPage) page;
            String baseCategoryPath;

            if(EdenUtils.isEmpty(post.getCategory())) {
                baseCategoryPath = "posts";
            }
            else {
                baseCategoryPath = "posts/" + post.getCategory();
            }

            String formattedFilename = PostsUtils.getPostFilename(post.getResource(), baseCategoryPath);

            Matcher matcher = PostsGenerator.pageTitleRegex.matcher(formattedFilename);

            if (matcher.matches()) {
                return matcher.group(4);
            }
        }
        else if(page instanceof PostArchivePage) {
            return ((PostArchivePage) page).getCategory();
        }

        return null;
    }

}

File: plugins/OrchidPosts/src/main/kotlin/com/eden/orchid/posts/permalink/pathTypes/SlugPathType.kt
package com.eden.orchid.posts.permalink.pathTypes

import com.eden.common.util.EdenUtils
import com.eden.orchid.api.theme.pages.OrchidPage
import com.eden.orchid.posts.PostsGenerator
import com.eden.orchid.posts.pages.PostArchivePage
import com.eden.orchid.posts.pages.PostPage
import com.eden.orchid.posts.permalink.PermalinkPathType
import com.eden.orchid.posts.utils.PostsUtils
import javax.inject.Inject

class SlugPathType @Inject
constructor() : PermalinkPathType(100) {

    override fun acceptsKey(page: OrchidPage, key: String): Boolean {
        return key == "slug"
    }

    override fun format(page: OrchidPage, key: String): String? {
        if (page is PostPage) {
            val baseCategoryPath: String

            if (EdenUtils.isEmpty(page.category)) {
                baseCategoryPath = "posts"
            } else {
                baseCategoryPath = "posts/" + page.category!!
            }

            val formattedFilename = PostsUtils.getPostFilename(page.resource, baseCategoryPath)

            val matcher = PostsGenerator.pageTitleRegex.matcher(formattedFilename)

            if (matcher.matches()) {
                return matcher.group(4)
            }
        } else if (page is PostArchivePage) {
            return page.category
        }

        return null
    }

}

